core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -590,10 +590,24 @@ private[spark] class BlockManager(
   private def doGetRemote(blockId: BlockId, asBlockResult: Boolean): Option[Any] = {
     require(blockId != null, "BlockId is null")
     val locations = Random.shuffle(master.getLocations(blockId))
+    var attemptTimes = 0
     for (loc <- locations) {
       logDebug(s"Getting remote block $blockId from $loc")
-      val data = blockTransferService.fetchBlockSync(
-        loc.host, loc.port, loc.executorId, blockId.toString).nioByteBuffer()
+      val data = try {
+        blockTransferService.fetchBlockSync(
+          loc.host, loc.port, loc.executorId, blockId.toString).nioByteBuffer()
+      } catch {
+        case t: Throwable if attemptTimes < locations.size - 1 =>
+          // Return null when Exception throw, so we can fetch block
+          // from another location if there still have locations
+          attemptTimes += 1
+          logWarning(s"Try $attemptTimes times getting remote block $blockId from $loc failed.", t)
+          null
+        case t: Throwable =>
+          // Throw BlockFetchException wraps the last Exception when
+          // there is no block we can fetch
+          throw new BlockFetchException(t)
```

Comment:
can we include a little more info here, eg. `s"failed to fetch block from ${locations.size} locations.  Most recent failure cause:"`